### PR TITLE
Add balloon pfn compact property based tests

### PR DESCRIPTION
**Pull request by sandreim**
# Reason for This PR

#2480

Adds 2 new property based tests for compact_page_frame_numbers()

## Description of Changes

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/devices/src/virtio/balloon/utils.rs [hunk 1]
```diff
@@ -270,4 +270,58 @@ mod tests {
             RemoveRegionError::MmapFail(_)
         );
     }
+
+    /// -------------------------------------
+    /// BEGIN PROPERTY BASED TESTING 
+
+    use proptest::prelude::*;
+
+    fn random_pfn_16k_u32_max() -> impl Strategy<Value = Vec<u32>> {
```
Comment (georgepisaltu) [inline src/devices/src/virtio/balloon/utils.rs:279]: Why do these functions have an array size of 16K? The PFN buffer has a size of 2K, as the `MAX_PAGE_COMPACT_BUFFER` constant suggests. Shouldn't we test with this constant instead?

**Review by georgepisaltu — COMMENTED** [review]
The tests look good.

As this is a completely new thing for the team, I would like us to provide the best documentation and code comments possible here in order to provide an example for future property testing in Firecracker. My comments should reflect this sentiment.

File: src/devices/src/virtio/balloon/utils.rs [hunk 2]
```diff
@@ -270,4 +270,58 @@ mod tests {
...
+            }
+        }
+
+        result
+    }
+
+    fn sort_and_dedup<T: Ord + Clone>(v: &[T]) -> Vec<T> {
+        let mut sorted_v = v.to_vec();
+        sorted_v.sort_unstable();
+        sorted_v.dedup();
```
[the remark this answers is not in the document]
Comment (sandreim) [reply src/devices/src/virtio/balloon/utils.rs:305]: I think this assumption holds true for test code in general.

File: src/devices/src/virtio/balloon/utils.rs [hunk 3]
```diff
@@ -270,4 +270,58 @@ mod tests {
...
+    proptest! {
+        #[test]
+        fn test_pfn_compact_1(mut input in random_pfn_16k_u32_max()) {
+            // The uncompactor will output sorted elements.
+            let sorted_input = sort_and_dedup(input.as_slice());
+            prop_assert!(uncompact(compact_page_frame_numbers(input.as_mut_slice())) == sorted_input);
+        }
+
+        #[test]
+        fn test_pfn_compact_2(mut input in random_pfn_16k_100()) {
```
[the remark this answers is not in the document]
Comment (sandreim) [reply src/devices/src/virtio/balloon/utils.rs:322]: The second test enforces duplicates PFNs. I'll add a comment.

**Comment by sandreim** [comment]
Updated. @georgepisaltu PTAL.

File: src/devices/src/virtio/balloon/utils.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply src/devices/src/virtio/balloon/utils.rs:304]: Fair enough.

File: src/devices/src/virtio/balloon/utils.rs [hunk 4]
```diff
@@ -270,4 +270,58 @@ mod tests {
             RemoveRegionError::MmapFail(_)
         );
     }
+
+    /// -------------------------------------
+    /// BEGIN PROPERTY BASED TESTING
+    use proptest::prelude::*;
+
+    fn random_pfn_u32_max() -> impl Strategy<Value = Vec<u32>> {
+        // Create a randomly sized vec (max 16k) filled with random u32 elements.
```
Comment (georgepisaltu) [inline src/devices/src/virtio/balloon/utils.rs:279]: Update the comment with the new size (`MAX_PAGE_COMPACT_BUFFER`).

The next comment as well.

Comment (sandreim) [reply src/devices/src/virtio/balloon/utils.rs]: done

**Review by georgepisaltu — APPROVED** [review]

File: src/devices/src/virtio/balloon/utils.rs [hunk 5]
```diff
@@ -270,4 +270,58 @@ mod tests {
...
+
+    // The uncompactor will output deduplicated and sorted elements as compaction algorithm
+    // guarantees it.
+    fn uncompact(compacted: Vec<(u32, u32)>) -> Vec<u32> {
+        let mut result = Vec::new();
+        for element in compacted.iter() {
+            for i in 0..element.1 {
+                result.push(element.0 + i);
+            }
+        }
```
Comment (nchong-at-aws) [inline src/devices/src/virtio/balloon/utils.rs:296]: Minor: could rephrase as
```
    for (start, len) in compacted {
        result.extend(start..=(start + len - 1));
    }
```

File: src/devices/src/virtio/balloon/utils.rs [hunk 6]
```diff
@@ -270,4 +270,58 @@ mod tests {
...
+    // The below prop tests will validate the following output proprties:
+    // - vec elements are sorted by first tuple value
+    // - no pfn duplicates are present
+    // - no pfn is lost
+    proptest! {
+        #[test]
+        fn test_pfn_compact_1(mut input in random_pfn_u32_max()) {
+            // The uncompactor will output sorted elements.
+            let sorted_input = sort_and_dedup(input.as_slice());
+            prop_assert!(uncompact(compact_page_frame_numbers(input.as_mut_slice())) == sorted_input);
```
Comment (nchong-at-aws) [inline src/devices/src/virtio/balloon/utils.rs:317]: Minor: could lift these into a single function invoked by both tests

**Review by nchong-at-aws — APPROVED** [review]
Looks great. 2 minor comments.

File: src/devices/src/virtio/balloon/utils.rs [hunk 7]
```diff
@@ -270,4 +270,58 @@ mod tests {
...
+    }
+
+    fn sort_and_dedup<T: Ord + Clone>(v: &[T]) -> Vec<T> {
+        let mut sorted_v = v.to_vec();
+        sorted_v.sort_unstable();
+        sorted_v.dedup();
+        sorted_v
+    }
+
+    // The below prop tests will validate the following output proprties:
```
Comment (luminitavoicu) [inline src/devices/src/virtio/balloon/utils.rs:308]:
```suggestion
    // The below prop tests will validate the following output properties:
```

File: src/devices/src/virtio/balloon/utils.rs (hunk 5 above)
Comment (sandreim) [reply src/devices/src/virtio/balloon/utils.rs:296]: done

File: src/devices/src/virtio/balloon/utils.rs (hunk 6 above)
Comment (sandreim) [reply src/devices/src/virtio/balloon/utils.rs:317]: Refactored into 1 test.

**Review by georgepisaltu — APPROVED** [review]

**Review by luminitavoicu — COMMENTED** [review]
LGTM

**Review by luminitavoicu — APPROVED** [review]

**Pull request merged**